_posts/2018-09-26-Inter-service-communication.md

```diff
@@ -0,0 +1,216 @@
+    <dependency>
+        <groupId>org.springframework.cloud</groupId>
+        <artifactId>spring-cloud-starter-openfeign</artifactId>
+    </dependency>
+{% endhighlight %}
+
+When we inspect the dependency module, we see that there is a lot coming out-of-the-box with the Spring Cloud Starter.
+If you are providing your own resilience or load balancing library you can just add the necessary dependencies you need.
+Be aware that the syntax is different between using the Spring wrapper or OpenFeign itself.
+To let your Spring context know that it has to search for Feign clients, we just add `@EnableFeignClients` on top of the main configuration class. 
+
+# Feign Clients 
+After we've enabled Feign on our classpath, we can start adding Feign clients. 
+When defining our clients, we have two solutions you can choose from. 
+The OpenFeign library, which provides us with the basic but very customizable Feign clients, and the Spring library, that adds a few extra libraries to it for cloud tooling.
+
+## Spring
+
+{% highlight java %}
+    @FeignClient(value = "auth", fallback = FallbackAuthClient.class, configuration = FeignConfig.class)
+    public interface AuthClient { 
+        @RequestMapping(method = RequestMethod.GET, value = "checkToken")
+        boolean isTokenValid(@RequestHeader("Authorization") String token);
+    }
+{% endhighlight %}
+
+Explanation properties:
+
+* `@FeignClient`: is the annotation for Spring to recognize Feign clients, Feign clients have to be interfaces as it is self declarative.
+* `value/name`: is the name of the application you are communicating with, Service Discovery ( when used ) uses this to bind them to the correct IP
+* `fallback`: is the Hystrix circuit breaker that comes out-of-the-box. Just make a new class, implement the interface and provide a proper fallback. 
+
+{% highlight java %}
+    @Component
+    public class FallbackAuthClient implements AuthClient {
+        @Override
+        public boolean isTokenValid(@RequestHeader("Authorization") String token) {
+            return false;
+        }
+    }
+{% endhighlight %}
+
+* `configuration`: is for extra configuration like logging, interceptors, etc... more on that below
+* `@RequestMapping`: inherits the Spring way of defining a request
+
+## OpenFeign
+For these clients we need an extra configuration class to tell Spring that this is the Feign client.
+
+{% highlight java %}
+
+public interface AuthClient {
+    @RequestLine("GET /auth")
+    @Headers({"Authorization: {token}", "Accept: application/hal+json"})
+    boolean isValid(@Param("token") String token);
+
+{% endhighlight %}
+* `@RequestLine`: is defining which verb and which uri path you are communicating to. 
+* `@Headers`: is defining the request headers that come with the request
+
+# Building an advanced custom Feign client
+When we need more advanced tooling inside our Feign client, there is the Feign configuration class where we can build and customize our Feign client. 
```

Comment:
> there is the Feign configuration class where we can build and customize our Feign client.

Are you talking about the `Feign.Builder` class here, or are you talking about a class which is annotated with `@Configuration` so we can customize our Feign client?